libnetwork/osl/interface_linux.go

```diff
@@ -276,22 +292,217 @@ func (n *Namespace) AddInterface(ctx context.Context, srcName, dstPrefix string,
+					// already have called the ErrorCallback, so the issue has been logged.
+					return
+				}
+				if u.Attrs().Index != ifIndex {
+					continue
+				}
+				log.G(ctx).WithFields(log.Fields{
+					"iface": u.Attrs().Name,
+					"ifi":   u.Attrs().Index,
+					"flags": deviceFlags(u.Flags),
+				}).Debug("link update")
+				if u.Flags&unix.IFF_NO_CARRIER != 0 {
+					// The link will never be ready.
+					return
+				}
+				if u.Flags&wantedFlags == wantedFlags {
+					return
+				}
+			}
+		}
+	}()
+
+	return upped, nil
+}
+
+// advertiseAddrs triggers send gratuitous ARP and Neighbour Advertisement
+// messages, so that caches are updated with the MAC address currently associated
+// with the interface's IP addresses.
+//
+// IP addresses are recycled quickly when endpoints are dropped on network
+// disconnect or container stop. A new MAC address may have been generated, so
+// this is necessary to avoid packets sent to the old MAC address getting dropped
+// until the ARP/Neighbour cache entries expire.
+//
+// Note that the kernel's arp_notify sysctl setting is not respected.
+func (n *Namespace) advertiseAddrs(ctx context.Context, ifIndex int, i *Interface, nlh nlwrap.Handle) {
+	macStr := i.MacAddress().String()
+	logFields := log.Fields{
+		"iface": i.dstName,
+		"ifi":   ifIndex,
+		"mac":   macStr,
+		"ip4":   i.Address(),
+		"ip6":   i.AddressIPv6(),
+	}
+	ctx = log.WithLogger(ctx, log.G(ctx).WithFields(logFields))
+
+	if i.address == nil && i.addressIPv6 == nil {
+		log.G(ctx).Debug("No IP addresses to advertise")
+		return
+	}
+
+	arpSender, naSender := n.prepAdvertiseAddrs(ctx, i, ifIndex)
+	if arpSender == nil && naSender == nil {
+		return
+	}
+	cleanup := func() {
+		if arpSender != nil {
+			arpSender.Close()
+		}
+		if naSender != nil {
+			naSender.Close()
+		}
+	}
+	stillSending := false
+	defer func() {
+		if !stillSending {
+			cleanup()
+		}
+	}()
+
+	send := func(ctx context.Context, nSent int) error {
+		uppedIface, err := nlh.LinkByIndex(ifIndex)
+		if err != nil {
+			log.G(ctx).WithError(err).Warn("failed to refresh link attributes")
+			return err
+		}
+		if curMAC := uppedIface.Attrs().HardwareAddr.String(); curMAC != macStr {
+			log.G(ctx).WithFields(log.Fields{"newMAC": curMAC}).Warn("MAC address changed")
+			return fmt.Errorf("MAC address changed, got %s, expected %s", curMAC, macStr)
+		}
+		log.G(ctx).WithFields(log.Fields{"n": nSent + 1}).Debug("Sending unsolicited ARP/NA")
+		var errs []error
+		if arpSender != nil {
+			if err := arpSender.Send(); err != nil {
+				log.G(ctx).WithError(err).Warn("Failed to send unsolicited ARP")
+				errs = append(errs, err)
+			}
+		}
+		if naSender != nil {
+			if err := naSender.Send(); err != nil {
+				log.G(ctx).WithError(err).Warn("Failed to send unsolicited NA")
+				errs = append(errs, err)
+			}
+		}
+		return stderrors.Join(errs...)
+	}
+
+	// Send an initial message. If it fails, skip the resends.
+	if send(ctx, 0) != nil {
+		return
```

Comment:
If the initial unsolicited ARP/NA messages can't be sent, should we log the error, or even bail out?

Reply:
It'll be logged by `send` itself ... but I figured it'd be best not to treat errors here as fatal - just in case there's some host/config where it never works, although that seems unlikely. (If it fails, the container will be up and running, and lack of ARP/NA won't always be an issue.)

But, now I realise setting the network's `advertise_addr_nmsgs` to zero would be a workaround if things do go wrong. So, yes - it returns the error now. Thank you.

Reply:
That found a bug ... Swarm doesn't set a MAC address in its `Interface`, so it wasn't sending anything.

Fixed by reading the address from the interface in that case, and checked that ARPs are now sent when starting a swarm service.

Reply:
Well, that was confusing! Not-sending an ARP/NA when the `Interface` doesn't have one was the right thing after all - in the swarm case, it was a bridge with no addresses moving netns and for its bridge/overlay networks, the ARPs were fine. L3-ipvlan endpoints also have no MAC address to advertise. So, just bail out when no MAC address is set up.

Reply:
Urgh, had to make another update ... I remembered `waitForIfUpped` waits for the interface to be `UP`+`RUNNING`+`LOWER_UP` - but, if it doesn't get there within 5s, container startup just carries on. (Before this PR, it didn't try to wait for the interface to report "up".)

Treating failure to send an ARP/NA as an error, rather than just logging, would unintentionally have made a "down" interface into an error.

So ... I got rid of the goroutine that waited for the interface to come up, made `waitForIfUpped` just block, and return a flag to say whether the interface is ready for unsolicited ARPs/NAs.

(The goroutine was there so to make it possible to subscribe to netlink updates before setting the interface "up". But, that hasn't been needed since I added `ListExisting` to the netlink options.)

That in turn made me realise the `NO_CARRIER` check was nonsense ... wrong flag! I got rid of it, but perhaps should bail out early if `LOWER_UP` is unset instead.